[DOC] Update TraceQL query editor doc (#2628)
grafana/tempo: docs/sources/tempo/traceql/query-editor.md

---

> title: Create TraceQL queries in Grafana
> menuTitle: Create TraceQL queries in Grafana

Author: ```suggestion
title: Write TraceQL queries in Grafana
menuTitle: Write TraceQL queries in Grafana
```

To add TraceQL panels to your dashboard, refer to the [Traces panel documentation](/docs/grafana/latest/panels-visualizations/visualizations/traces/).

{{% admonition type="note" %}}
To use the TraceQL query editor in Grafana 9.3.2 and newer, you need to enable the `traceqlEditor` feature flag.

To enable the Trace Search in Grafana 10 and newer, you need to enable the `traceqlSearch` feature flag.

The query editor and search are available in Grafana Cloud without enabling a feature flag.
{{% /admonition %}}

## Create TraceQL queries using the query editor

The Tempo data source’s query editor helps you query and display traces from Tempo in **Explore**.

To access the query editor, follow these steps:

1. Sign into Grafana or Grafana Cloud.
1. Select your Tempo data source.
1. From the menu, choose **Explore** and select the **TraceQL** tab.
1. Start your query on the text line by entering `{{`. For help with TraceQL syntax, refer to the [Construct a TraceQL query documentation]({{< relref "./_index.md" >}}).
1. Once you have finished your query, select **Run query**.

![Query editor showing request for http.method](/static/img/docs/tempo/query-editor-http-method.png)

> ### Query by TraceID

Reviewer: Can we add "Query by SpanID"? And also by time range? Maybe it's the selected time interval for the time range selection

Author: working on this today

Author: Okay, so query by Span ID does not work in the same way that Query by TraceID does. I'll figure out another medhot and write that up.

Author: I added an optional step to adjust the time range and linked to the documentation for the time picker.

Author: @cyrille-leclerc What do you think of the updates?

To query a particular trace:

1. From the menu, choose **Explore** and select the **TraceQL** tab.

> 1. Enter the trace’s ID into the query field.

Reviewer: Can we provide an example, is it between `{...}` brackets or without?

Author: I confirmed that adding a trace ID directly on the search field on the TraceQL tab and pressing Run query returns data for the trace ID. I have a screenshot I can add to this section, if you think that would be helpful.

  - TraceQL
---

> # Create TraceQL queries in Grafana

Reviewer: ```suggestion
# Write TraceQL queries in Grafana
```

"Create" sounds to me like something intended to last when here it's about exploring data, something not necessarily intended to be persisted.

Author: Thank you for the clarification.

> You can use compose TraceQL queries in Grafana and Grafana Cloud using the Search in Explore and the TraceQL viewer and query editor in the Tempo data source.

Reviewer: This sentence sounds to me like there is a "Tempo Datasource" screen that I can navigate to and write my TraceQL queries but I don't think that exists. It also suggest there is a view called "query editor", is it a reference to the trace query visual editor of the "Query Type" tab "Search"?

My understanding is that I have to
* Navigate to Explore
* Select a Tempo data source
* Navigate to the "TraceQL" tab. An alternative being to use the TraceQL visual builder of the "Search" tab.

Author: Thank you for the clarifications. I've reworded the content using your instructions. 

Query editor was the term used when we only had one location (TraceQL tab) to create TraceQL queries. I'll recheck the phrasing in this article ot make sure that the wording is correct. Thank you for the question.

> You can use compose TraceQL queries in Grafana and Grafana Cloud using **Explore** and a Tempo data source. You can use either the **Query type** > **Search** or the **TraceQL** tab.

Reviewer: I don't understand what _"You can use compose TraceQL queries in Grafana and Grafana Cloud"_ means. Should we remove "use"?

Author: It should be, You can compose. Thank you for the catch! 
```suggestion
You can compose TraceQL queries in Grafana and Grafana Cloud using **Explore** and a Tempo data source. You can use either the **Query type** > **Search** or the **TraceQL** tab.
```

Both of these methods let you build queries and drill-down into result sets.

To add TraceQL panels to your dashboard, refer to the [Traces panel documentation](/docs/grafana/latest/panels-visualizations/visualizations/traces/).

To learn more about Grafana dashboards, refer to the [Use dashboards documentation](/docs/grafana/latest/dashboards/use-dashboards/).

{{% admonition type="note" %}}

> To use the TraceQL query editor in Grafana 9.3.2 and newer, you need to enable the `traceqlEditor` feature flag.

Author: ```suggestion
To use the TraceQL query editor in Grafana 9.3.2 and newer, you need to enable the `traceqlEditor` feature toggle ([refer to instructions for enabling feature toggles](/docs/grafana/latest/setup-grafana/configure-grafana/feature-toggles/).)
```

> To enable the Trace Search in Grafana 10 and newer, you need to enable the `traceqlSearch` feature flag.

Reviewer: Comments that should not prevent from merging this PR:

1. I don't know where to edit Feature flags
2. Are we sue a feature flag must be changed? Are we sure TraceQL tab is not enabled by default in Grafana 10? I can see the TraceQL tab on my Grafana 10 on Grafana Cloud even though I don't remember tweaking my config

Author: I can add a link to the Feature Flag page easily enough. The feature flag isn't necessary in Grafana Cloud. i think both of these features are default in Cloud. But I will verify.

Reviewer: Thank you @knylander-grafana for your pugnacity.
I think it would be great, in this doc update or a subsequent one, to be clear about the need or not of a feature flag on Grafana Cloud, especially for Grafana 10

Reviewer 2: I agree with @cyrille-leclerc I don't think that `traceqlSearch` is required in Grafana 10.

@grafana/observability-traces-and-profiling do we have a list of what feature flags are required in which Grafana versions for the traceql search and the traceql builder replacement of old search?

Author: @adrapereira It's still confusing on this page that we have two panels on the same page that deal with TraceQL search. The Query type > Search is the TraceQL query builder while the TraceQL tab is the query editor. Using these terms is confusing -- especially if our own people are getting confused by the terminology. 

How do you refer to the editor and the builder?

Author: Here's what I did: 
![image](https://github.com/grafana/tempo/assets/104772500/e48c089f-92f2-4f62-b501-eb6945a4010c)

> To use the TraceQL query editor in Grafana 9.3.2 and newer, you need to enable the `traceqlEditor` feature toggle ([refer to instructions for enabling feature toggles](/docs/grafana/latest/setup-grafana/configure-grafana/feature-toggles/).)

Reviewer 2: It's not clear to me what the "TraceQL query editor" is. I think this is the search replacement with the TraceQL builder. Does this have an official name?

Author: Oh! Then I need to remove the feature toggle information and clarifyi the text.

Author: Would it be better to say the TraceQL tab instead of the query editor since we now also have a query builder in Query type > Search?

The query editor and search are available in Grafana Cloud without enabling a feature flag.
{{% /admonition %}}

> ## Create TraceQL queries using the query editor

Reviewer: ```suggestion
## Write TraceQL queries using the query editor
```

Additional question: what is the "Query Editor"? Is it the visual TraceQL builder of the "Query type" tab "Search"?

Author: This was a good point. We started off with only the TraceQL tab, so that was the TraceQL query editor. The Search tab was added on later, with the doc added to the same page. I've tried to clarify in this PR the differences. The query builder name is an excellent term to use to help clarify.

The Tempo data source’s query editor helps you query and display traces from Tempo in **Explore**.

To access the query editor, follow these steps:

1. Sign into Grafana or Grafana Cloud.
1. Select your Tempo data source.
1. From the menu, choose **Explore** and select the **TraceQL** tab.

> 1. Start your query on the text line by entering `{{`. For help with TraceQL syntax, refer to the [Construct a TraceQL query documentation]({{< relref "./_index.md" >}}).

Reviewer: ```suggestion
1. Start your query on the text line by entering `{`. For help with TraceQL syntax, refer to the [Construct a TraceQL query documentation]({{< relref "./_index.md" >}}).
```

Isn't it one single bracket? `{`

Author: Good catch. I verified that it is a single `{` and not two. Updated.

1. Optional: Use the Time picker drop-down to change the time and range for the query (refer to the [documentation for instructions](/docs/grafana/latest/dashboards/use-dashboards#set-dashboard-time-range)).
1. Once you have finished your query, select **Run query**.

![Query editor showing request for http.method](/static/img/docs/tempo/query-editor-http-method.png)

### Query by TraceID

> To query a particular trace:

Reviewer: ```suggestion
To query a particular trace by its trace ID:
```

> 1. From the menu, choose **Explore** and select the **TraceQL** tab.

Reviewer: ```suggestion
1. From the menu, choose **Explore**, select the desired Tempo datasource and navigate to the **TraceQL** tab.
```

Author: Done. I'm editing the file directly to make changes.

1. Enter the trace ID into the query field. For example: `1f187d8363b5a9b30cedd8e0ce9ccb43`

> 1. Select **Run query**.

Reviewer: ```suggestion
1. Click on **Run query** or use the keyboard shortcut shift+enter.
```

Author: I've updated the text. I tend to use Select instead of Click in case someone is using a touchscreen device like a tablet.
